Sources/Structs/Weak.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  Created by Murat Yilmaz on 19.05.18.
+//  Copyright © 2018 Flinesoft. All rights reserved.
+//
+
+import Foundation
+
+/// A wrapper for storing weak references to a `Wrapped` instance.
+public struct Weak<Wrapped>: ExpressibleByNilLiteral where Wrapped: AnyObject {
+    /// The value of `Wrapped` stored as weak reference
+    public weak var value: Wrapped?
+
+    /// Creates an instance that stores the given value.
+    public init(_ value: Wrapped) {
+        self.value = value
+    }
+
+    /// Evaluates the given closure when this `Weak` instance is not `nil`,
+    /// passing the value as a parameter.
+    ///
+    /// - Parameter transform: A closure that takes the unwrapped value
+    ///   of the instance.
+    /// - Returns: The result of the given closure. If this instance is `nil`,
+    ///   returns `nil`.
+    public func map<U>(_ transform: (Wrapped) throws -> U) rethrows -> U? {
+        guard let value = value else { return nil }
+
+        return try transform(value)
+    }
+
+    /// Evaluates the given closure when this `Weak` instance is not `nil`,
+    /// passing the value as a parameter.
+    ///
+    /// - Parameter transform: A closure that takes the unwrapped value
+    ///   of the instance.
+    /// - Returns: The result of the given closure. If this instance is `nil`,
+    ///   returns `nil`.
+    public func flatMap<U>(_ transform: (Wrapped) throws -> U?) rethrows -> U? {
+        guard let value = value else { return nil }
+
+        return try transform(value)
+    }
+
+    /// Creates an instance initialized with `nil`.
+    public init(nilLiteral: ()) {
+        self.value = nil
+    }
+
+    /// The wrapped value of this instance, unwrapped without checking whether
+    /// the instance is `nil`.
+    public var unsafelyUnwrapped: Wrapped {
```

Comment:
Do we really need this? I mean, the value is public, so it should be easy to do this already, or not?